Add --recursive option

DerZade opened the pull request:
Add `--recursive` option and therefore fix #42.

---

This PR includes two breaking changes to the public API:
- added second parameter (`recurse`) to `spreet::fs::get_svg_input_paths`
- added second parameter (`base_path`) to `spreet::sprite::sprite_name`

I hope this is acceptable. If not, just let me know and I will add other functions instead of extending the existing ones. But this will result in some duplicate code 🙈

flother commented on src/sprite.rs:
```diff
@@ -228,8 +228,18 @@ impl Spritesheet {
 /// Returns the name (unique id within a spritesheet) taken from a file.
-pub fn sprite_name(path: &Path) -> String {
-    format!("{}", path.file_stem().unwrap().to_string_lossy())
+pub fn sprite_name(path: &Path, base_path: &Path) -> String {
```
Now it's doing something a little more complicated, I think this function's a perfect candidate for a unit test or two (I'm holding you to a higher standard than I hold myself, sorry!) Do you fancy adding some? One for when base_path has no impact, one when it does

DerZade replied:
Addressed in 73fda65003dc95ce4a639e30c65a1ad57d4d6aa4

flother commented on src/sprite.rs:
```diff
@@ -228,8 +228,18 @@ impl Spritesheet {
 }
 
 /// Returns the name (unique id within a spritesheet) taken from a file.
```
Let's add an explanation of what `base_path` is used for

```suggestion
/// Returns the name (unique id within a spritesheet) taken from a file.
///
/// If `base_path` is given, it's used as prefix in the sprite name (minus any parents
/// shared with `path`).
```

DerZade replied:
You're explanation starts with _"If `base_path` is given"_. There is not really a choice tho 😂 so I adjusted it a bit. 

Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

flother reviewed (commented):
This is great, I appreciate you taking the time to add this in. I have a few comments on minor things I'd like to see, but I'll be happy to merge this.

Other than the comments below, could you add a line to [`CHANGELOG.md`](https://github.com/flother/spreet/blob/master/CHANGELOG.md) to summarise the new flag?

DerZade commented on README.md:
```diff
@@ -99,6 +99,7 @@ Options:
   -r, --ratio <RATIO>      Set the output pixel ratio [default: 1]
       --retina             Set the pixel ratio to 2 (equivalent to `--ratio=2`)
       --unique             Store only unique images in the spritesheet, and map them to multiple names
+      --recursive          Include images from subdirectories
```
Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

DerZade commented on src/fs.rs:
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
-/// (files whose names begin with `.`) but it does follow symlinks.
-pub fn get_svg_input_paths(path: &Path) -> Vec<PathBuf> {
+/// It ignores hidden files (files whose names begin with `.`) but it does follow symlinks.
+pub fn get_svg_input_paths(path: &Path, recurse: bool) -> Vec<PathBuf> {
```
Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

DerZade commented on src/fs.rs:
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
+            let path_buf = entry.path();
+
+            if recurse && path_buf.is_dir() {
+                return Some(get_svg_input_paths(&path_buf.as_path(), recurse));
```
Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

DerZade commented on src/fs.rs:
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
+            if recurse && path_buf.is_dir() {
+                return Some(get_svg_input_paths(&path_buf.as_path(), recurse));
+            }
+
```
Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

DerZade commented on src/fs.rs:
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
+            if is_useful_input(&entry) {
+                return Some(vec![path_buf]);
+            }
+
```
Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

DerZade commented:
I have addressed most of your comments and have added a line to the change log.

The only thing left to do is https://github.com/flother/spreet/pull/43#discussion_r1103876825, which I hope to get to tonight :)

DerZade commented:
I added tests for `sprite_name` in 73fda65003dc95ce4a639e30c65a1ad57d4d6aa4 🙃

flother commented on src/fs.rs:
```diff
@@ -31,14 +31,23 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
-        .map(|entry| entry.path())
+        .filter_map(|entry| {
+            let path_buf = entry.path();
+
+            if recursive && path_buf.is_dir() {
+                return Some(get_svg_input_paths(path_buf.as_path(), recursive));
+            }
+            if is_useful_input(&entry) {
+                return Some(vec![path_buf]);
+            }
+            return None;
```
```suggestion
            if recursive && path_buf.is_dir() {
                Some(get_svg_input_paths(path_buf.as_path(), recursive))
            } else if is_useful_input(&entry) {
                Some(vec![path_buf])
            } else {
                None
            }
```

flother reviewed (approved):
Brilliant, thanks for taking the time to add this

DerZade commented:
Thanks a lot for the most pleasant OSS experience I had in a long time :)
